File: 02/main.py
#!/usr/bin/env python3
from functools import reduce

with open('data.txt') as f:
    data = [x.split() for x in f.read().split('\n') if x != '']

position = sum(int(xy[1]) for xy in data if xy[0] == 'forward')
depth = sum(int(xy[1]) if xy[0] == 'down' else -int(xy[1]) for xy in data if xy[0] != 'forward')
print(position*depth)

position, depth, aim = 0, 0, 0
for cmd, x in data:
    if cmd == 'down':
        aim += int(x)
    elif cmd == 'up':
        aim -= int(x)
    else:
        position += int(x)
        depth += aim*int(x)

print(position*depth) 

File: 02/main.hs
(|>) :: (a -> b) -> (b -> c) -> a -> c
f |> g = g . f

data CMD  = Forward Int | Up Int | Down Int

parse :: [String] -> CMD
parse (cmd:y:_) = case cmd of
    "forward" -> Forward x
    "up" -> Up x
    "down" -> Down x
    _ -> error "Invalid Input"
    where x = read y

parse _ = error "Invalid Input"

f :: [CMD] -> Int
f = foldl get (0,0) |> uncurry (*)
    where
        get (pos, depth) cmd = case cmd of 
            Forward x -> (pos + x, depth)
            Up x -> (pos, depth - x)
            Down x -> (pos, depth + x)

g :: [CMD] -> Int
g = foldl get (0,0,0) |> \(a,b,_) -> a*b
    where
        get (pos, depth, aim) cmd = case cmd of 
            Forward x -> (pos + x, depth + aim*x, aim)
            Up x      -> (pos, depth, aim - x)
            Down x    -> (pos, depth, aim + x)


main :: IO ()
main = interact $ 
    lines |> map (words |> parse) |> (\xs -> [f xs, g xs]) |> map show |> unlines